src/sage/rings/valuation/inductive_valuation.py

```diff
@@ -0,0 +1,1526 @@
+
+                w_coefficients = w.coefficients(G)
+                if principal_part_bound:
+                    w_coefficients = islice(w_coefficients, 0, principal_part_bound + 1, 1)
+                w_coefficients = list(w_coefficients)
+
+                w_valuations = w.valuations(G, coefficients=w_coefficients)
+                if principal_part_bound:
+                    w_valuations = islice(w_valuations, 0, principal_part_bound + 1, 1)
+                w_valuations = list(w_valuations)
+
+                from sage.geometry.newton_polygon import NewtonPolygon
+                NP = NewtonPolygon(w.newton_polygon(G, valuations=w_valuations).vertices(), last_slope=0)
+
+                verbose("Newton-Polygon for v(phi)=%s : %s"%(self(phi), NP), level=11)
+                slopes = NP.slopes(repetition=True)
+                multiplicities = {slope : len([s for s in slopes if s == slope]) for slope in slopes}
+                slopes = multiplicities.keys()
+                if NP.vertices()[0][0] != 0:
+                    slopes = [-infinity] + slopes
+                    multiplicities[-infinity] = 1
+
+                if not slopes:
+                    q,r = G.quo_rem(phi)
+                    assert not r.is_zero()
+                    phi = phi.coefficients(sparse=False)
+                    for i,c in enumerate(r.coefficients(sparse=False)):
+                        if not c.is_zero():
+                            v = w(c)
+                            # for a correct result we need to add O(pi^v) in degree i
+                            # we try to find the coefficient of phi where such an
+                            # error can be introduced without losing much absolute
+                            # precision on phi
+                            best = i
+                            for j in range(i):
+                                if w(q[j]) < w(q[best]):
+                                    best = j
+                            # now add the right O() to phi in degree i - best
+                            phi[i-best] = phi[i-best].add_bigoh(w(c)-w(q[best]))
+
+                    phi = G.parent()(phi)
+                    w = self._base_valuation.augmentation(phi, infinity, check=False)
+                    ret.append((w, phi.degree(), principal_part_bound, None, None))
+                    continue
+
+                for i, slope in enumerate(slopes):
+                    slope = slopes[i]
+                    verbose("Slope = %s"%slope, level=12)
+                    new_mu = old_mu - slope
+                    new_valuations = [val - (j*slope if slope is not -infinity else (0 if j == 0 else -infinity)) for j,val in enumerate(w_valuations)]
+                    base = self
+                    if phi.degree() == base.phi().degree():
+                        assert new_mu > self(phi)
+                        if not base.is_gauss_valuation():
+                            base = base._base_valuation
```

Comment:
Is it ever possible to have two augmentations in a row with the same degree?  In that case, is this the right thing to do?